[BE][FSDP] Introduce `FSDPTestModel` interface

awgu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #81158
* #80915
* #80908
* __->__ #80873

**Overview**
Please refer to https://github.com/pytorch/pytorch/issues/80867 first.

This addresses:
> Goal 3: Refactor model construction to enable simpler testing for the non-recursive wrapping path.

The idea is that we have an abstract class `FSDPTestModel` that defines the interface expected from the parity check and training boilerplate. This PR refactors the models in `common_fsdp.py` used in `test_fsdp_core.py` to implement this interface. Further unification under this interface is coming in follow-up PRs.

Commit:
[BE][FSDP] Introduce `FSDPTestModel` interface

[ghstack-poisoned]

Commit:
Update on "[BE][FSDP] Introduce `FSDPTestModel` interface"


**Overview**
Please refer to https://github.com/pytorch/pytorch/issues/80867 first.

This addresses:
> Goal 3: Refactor model construction to enable simpler testing for the non-recursive wrapping path.

**Notes**


[ghstack-poisoned]

Commit:
Update on "[BE][FSDP] Introduce `FSDPTestModel` interface"


**Overview**
Please refer to https://github.com/pytorch/pytorch/issues/80867 first.

This addresses:
> Goal 3: Refactor model construction to enable simpler testing for the non-recursive wrapping path.

**Notes**


[ghstack-poisoned]

Commit:
Update on "[BE][FSDP] Introduce `FSDPTestModel` interface"


**Overview**
Please refer to https://github.com/pytorch/pytorch/issues/80867 first.

This addresses:
> Goal 3: Refactor model construction to enable simpler testing for the non-recursive wrapping path.

**Notes**


[ghstack-poisoned]

awgu commented on test/distributed/fsdp/test_fsdp_state_dict.py:
```diff
@@ -206,51 +224,62 @@ def test_fsdp_state_dict_with_activation_checkpoint(self, checkpoint_wrap):
         if self.rank == 0:
-            model_new.load_state_dict(state_dict)
-
-        # TransformerWithSharedParams has a buffer of zeros, so can't pass in
```
This was actually slightly incorrect. The issue is not just the `"long_buffer"` of zeros. All linear biases are initialized to zero by default. I added logic to force the model parameters and buffers to be nonzero above. This avoids this caveat altogether.

awgu commented on torch/testing/_internal/common_fsdp.py:
```diff
@@ -517,158 +836,141 @@ def _train_for_several_steps(
         except Exception as e:
-            raise ValueError(f"model_Init_fn {model_init_fn} got error {str(e)}")
-
-        cpu_offload = cpu_offload or CPUOffload()  # disabled if not specified.
```
I removed this to let the FSDP constructor take care of the default argument initialization instead of doing it in this method externally.

awgu commented on torch/testing/_internal/common_fsdp.py:
```diff
@@ -481,25 +799,26 @@ def _train_for_several_steps(
             if norm_type is not None:
-                if isinstance(model, FullyShardedDataParallel):
-                    model.clip_grad_norm_(clip_norm, norm_type)
+                max_norm = 0.3
```
I removed `max_norm` from the method arguments because no caller ever passes it, so it always takes its default value of 0.3. I also renamed to `max_norm` and `norm_type` to match the API.

awgu commented on torch/testing/_internal/common_fsdp.py:
```diff
@@ -172,43 +263,97 @@ def get_loss(self, input, output):
+        elif fsdp_init_mode == FSDPInitMode.RECURSIVE:
+            if "cpu_offload" in fsdp_kwargs and fsdp_kwargs["cpu_offload"].offload_params:
+                assert cuda_init_mode == CUDAInitMode.CUDA_NEVER, f"{cuda_init_mode}"
+            # Default to the `transformer_auto_wrap_policy()`
```
Now that we have the `transformer_auto_wrap_policy()`, it makes more sense to use this instead of always using a single top-level FSDP to wrap the `TransformerWithSharedParams`, which is not a meaningful wrapping.

awgu commented on torch/testing/_internal/common_fsdp.py:
```diff
@@ -104,28 +187,31 @@ def __init__(self, wrap_fsdp, cpu_offload=CPUOffload(offload_params=False)):
-        ), "dim of vocab should be larger than 12, as we use torch.arange(12) as input"
+        if deterministic:
+            torch.manual_seed(0)
+        d_vocab = 23
```
I removed these from the method arguments because no caller ever passes them, so they always take their default values of 23 and 16.

awgu commented on torch/testing/_internal/common_fsdp.py:
```diff
@@ -23,30 +33,97 @@
+    # FSDP recursive wrapping
+    RECURSIVE = auto()
+    # TODO: FSDP non-recursive wrapping
+    # NONRECURSIVE = auto()
```
The idea is that when I am testing the non-recursive wrapping, I implement a branch for `fsdp_init_mode == FSDPInitMode.NONRECURSIVE` for each model in `common_fsdp.py` and then pass `fsdp_init_mode=FSDPInitMode.NONRECURSIVE` from each existing test. There should not need to be any other changes.

awgu commented on test/distributed/fsdp/test_wrap.py:
```diff
@@ -365,23 +366,28 @@ def test_always_wrap(self):
         model = FSDP(seq, process_group=self.process_group, auto_wrap_policy=always_wrap_policy)
         TestFSDPWrap.NestedSequentialModel.verify_model_all_wrapped(self, model)
 
-    @unittest.skipIf(torch.cuda.device_count() < 2, "Requires at least 2 GPUs")
```
We should make all of these consistent and use `skip_if_lt_x_gpu()` as a follow-up.

awgu commented on test/distributed/fsdp/test_fsdp_sharded_grad_scaler.py:
```diff
@@ -3,24 +3,32 @@
 import itertools
 import sys
-import torch
 import unittest
```
Sorted imports (for all test files touched).

Commit:
Update on "[BE][FSDP] Introduce `FSDPTestModel` interface"


**Overview**
Please refer to https://github.com/pytorch/pytorch/issues/80867 first.

This addresses:
> Goal 3: Refactor model construction to enable simpler testing for the non-recursive wrapping path.

**Notes**


[ghstack-poisoned]

awgu commented on test/distributed/fsdp/test_fsdp_core.py:
```diff
@@ -105,138 +114,177 @@ def test_nested_wrapped_model(self, cpu_offload, backward_prefetch, forward_pref
     @skip_if_lt_x_gpu(2)
     @parametrize(params, configs, subtest_name)
-    @parametrize("clip_norm_type", [2.0, None])
+    # TODO (awgu): 2.0 fails tests
```
https://github.com/pytorch/pytorch/issues/80872

awgu reviewed (commented):
I am leaving a few comments to help review.

Commit:
Update on "[BE][FSDP] Introduce `FSDPTestModel` interface"


**Overview**
Please refer to https://github.com/pytorch/pytorch/issues/80867 first.

This addresses:
> Goal 3: Refactor model construction to enable simpler testing for the non-recursive wrapping path.

The idea is that we have an abstract class `FSDPTestModel` that defines the interface expected from the parity check and training boilerplate. This PR refactors the models in `common_fsdp.py` used in `test_fsdp_core.py` to implement this interface. Further unification under this interface is coming in follow-up PRs.


[ghstack-poisoned]

Commit:
Update on "[BE][FSDP] Introduce `FSDPTestModel` interface"


**Overview**
Please refer to https://github.com/pytorch/pytorch/issues/80867 first.

This addresses:
> Goal 3: Refactor model construction to enable simpler testing for the non-recursive wrapping path.

The idea is that we have an abstract class `FSDPTestModel` that defines the interface expected from the parity check and training boilerplate. This PR refactors the models in `common_fsdp.py` used in `test_fsdp_core.py` to implement this interface. Further unification under this interface is coming in follow-up PRs.


[ghstack-poisoned]

Commit:
Update on "[BE][FSDP] Introduce `FSDPTestModel` interface"


**Overview**
Please refer to https://github.com/pytorch/pytorch/issues/80867 first.

This addresses:
> Goal 3: Refactor model construction to enable simpler testing for the non-recursive wrapping path.

The idea is that we have an abstract class `FSDPTestModel` that defines the interface expected from the parity check and training boilerplate. This PR refactors the models in `common_fsdp.py` used in `test_fsdp_core.py` to implement this interface. Further unification under this interface is coming in follow-up PRs.


[ghstack-poisoned]

Commit:
Update on "[BE][FSDP] Introduce `FSDPTestModel` interface"


**Overview**
Please refer to https://github.com/pytorch/pytorch/issues/80867 first.

This addresses:
> Goal 3: Refactor model construction to enable simpler testing for the non-recursive wrapping path.

The idea is that we have an abstract class `FSDPTestModel` that defines the interface expected from the parity check and training boilerplate. This PR refactors the models in `common_fsdp.py` used in `test_fsdp_core.py` to implement this interface. Further unification under this interface is coming in follow-up PRs.


[ghstack-poisoned]

Commit:
Update on "[BE][FSDP] Introduce `FSDPTestModel` interface"


**Overview**
Please refer to https://github.com/pytorch/pytorch/issues/80867 first.

This addresses:
> Goal 3: Refactor model construction to enable simpler testing for the non-recursive wrapping path.

The idea is that we have an abstract class `FSDPTestModel` that defines the interface expected from the parity check and training boilerplate. This PR refactors the models in `common_fsdp.py` used in `test_fsdp_core.py` to implement this interface. Further unification under this interface is coming in follow-up PRs.


[ghstack-poisoned]

Commit:
Update on "[BE][FSDP] Introduce `FSDPTestModel` interface"


**Overview**
Please refer to https://github.com/pytorch/pytorch/issues/80867 first.

This addresses:
> Goal 3: Refactor model construction to enable simpler testing for the non-recursive wrapping path.

The idea is that we have an abstract class `FSDPTestModel` that defines the interface expected from the parity check and training boilerplate. This PR refactors the models in `common_fsdp.py` used in `test_fsdp_core.py` to implement this interface. Further unification under this interface is coming in follow-up PRs.


[ghstack-poisoned]

Commit:
Update on "[BE][FSDP] Introduce `FSDPTestModel` interface"


**Overview**
Please refer to https://github.com/pytorch/pytorch/issues/80867 first.

This addresses:
> Goal 3: Refactor model construction to enable simpler testing for the non-recursive wrapping path.

The idea is that we have an abstract class `FSDPTestModel` that defines the interface expected from the parity check and training boilerplate. This PR refactors the models in `common_fsdp.py` used in `test_fsdp_core.py` to implement this interface. Further unification under this interface is coming in follow-up PRs.


[ghstack-poisoned]

Commit:
Update on "[BE][FSDP] Introduce `FSDPTestModel` interface"


**Overview**
Please refer to https://github.com/pytorch/pytorch/issues/80867 first.

This addresses:
> Goal 3: Refactor model construction to enable simpler testing for the non-recursive wrapping path.

The idea is that we have an abstract class `FSDPTestModel` that defines the interface expected from the parity check and training boilerplate. This PR refactors the models in `common_fsdp.py` used in `test_fsdp_core.py` to implement this interface. Further unification under this interface is coming in follow-up PRs.


[ghstack-poisoned]

Commit:
Update on "[BE][FSDP] Introduce `FSDPTestModel` interface"


**Overview**
Please refer to https://github.com/pytorch/pytorch/issues/80867 first.

This addresses:
> Goal 3: Refactor model construction to enable simpler testing for the non-recursive wrapping path.

The idea is that we have an abstract class `FSDPTestModel` that defines the interface expected from the parity check and training boilerplate. This PR refactors the models in `common_fsdp.py` used in `test_fsdp_core.py` to implement this interface. Further unification under this interface is coming in follow-up PRs.


[ghstack-poisoned]

awgu commented on test/distributed/fsdp/test_fsdp_misc.py:
```diff
@@ -51,111 +50,121 @@ def process_group(self):
 
-        def _check_device_matches(fsdp, dev_id):
-            devices = {p.device for p in fsdp.parameters()}
+        def _check_device_matches(module, device_id):
```
Changed this argument to `device_id` just to prevent accidental variable shadowing.

awgu commented on torch/testing/_internal/common_fsdp.py:
```diff
@@ -23,30 +33,97 @@
-    CUDA_NEVER = 3
-
-
-def _validate(model, process_group, assert_fn):
```
Changed this to `_assert_module_states()` since `_validate()` was a bit non-descriptive.

Commit:
Update on "[BE][FSDP] Introduce `FSDPTestModel` interface"


**Overview**
Please refer to https://github.com/pytorch/pytorch/issues/80867 first.

This addresses:
> Goal 3: Refactor model construction to enable simpler testing for the non-recursive wrapping path.

The idea is that we have an abstract class `FSDPTestModel` that defines the interface expected from the parity check and training boilerplate. This PR refactors the models in `common_fsdp.py` used in `test_fsdp_core.py` to implement this interface. Further unification under this interface is coming in follow-up PRs.


[ghstack-poisoned]

Commit:
Update on "[BE][FSDP] Introduce `FSDPTestModel` interface"


**Overview**
Please refer to https://github.com/pytorch/pytorch/issues/80867 first.

This addresses:
> Goal 3: Refactor model construction to enable simpler testing for the non-recursive wrapping path.

The idea is that we have an abstract class `FSDPTestModel` that defines the interface expected from the parity check and training boilerplate. This PR refactors the models in `common_fsdp.py` used in `test_fsdp_core.py` to implement this interface. Further unification under this interface is coming in follow-up PRs.


[ghstack-poisoned]

Commit:
Update on "[BE][FSDP] Introduce `FSDPTestModel` interface"


**Overview**
Please refer to https://github.com/pytorch/pytorch/issues/80867 first.

This addresses:
> Goal 3: Refactor model construction to enable simpler testing for the non-recursive wrapping path.

The idea is that we have an abstract class `FSDPTestModel` that defines the interface expected from the parity check and training boilerplate. This PR refactors the models in `common_fsdp.py` used in `test_fsdp_core.py` to implement this interface. Further unification under this interface is coming in follow-up PRs.


[ghstack-poisoned]

awgu commented:
`test_mp_embedding_reduce()` fails if we wrap the `TransformerWithSharedParameters` with the typical `transformer_auto_wrap_policy()` with `transformer_layer_cls={TransformerEncoderLayer,TransformerDecoderLayer}`.

rohan-varma commented on test/distributed/fsdp/test_wrap.py:
```diff
@@ -367,21 +370,26 @@ def test_always_wrap(self):
+        encoder_layers = set(fsdp_model.module.transformer.encoder.layers)
+        decoder_layers = set(fsdp_model.module.transformer.decoder.layers)
+        for module in modules:
+            if module is fsdp_model or module in encoder_layers or module in decoder_layers:
```
maybe check `isinstance` to be more robust?

awgu replied:
I am not sure I follow the suggestion. `isinstance` is what we are checking below with the asserts, so it would not make sense to check it in the conditional.

rohan-varma commented on test/distributed/fsdp/test_wrap.py:
```diff
@@ -475,16 +483,16 @@ def test_auto_wrap_preset_force_leaf_custom(self):
 
     @unittest.skipIf(not torch.cuda.is_available(), "Test Requires CUDA")
-    @parametrize("fsdp_init_mode", [FSDPInitMode.CUDA_BEFORE, FSDPInitMode.CUDA_AFTER])
+    @parametrize("cuda_init_mode", [CUDAInitMode.CUDA_BEFORE, CUDAInitMode.CUDA_AFTER])
```
subTest?

awgu replied:
Sounds good. Subtesting will be done in follow ups.

rohan-varma reviewed (approved):
LGTM, maybe there can be some follow up work to use these models for all FSDP unittests.

Commit:
Update on "[BE][FSDP] Introduce `FSDPTestModel` interface"


**Overview**
Please refer to https://github.com/pytorch/pytorch/issues/80867 first.

This addresses:
> Goal 3: Refactor model construction to enable simpler testing for the non-recursive wrapping path.

The idea is that we have an abstract class `FSDPTestModel` that defines the interface expected from the parity check and training boilerplate. This PR refactors the models in `common_fsdp.py` used in `test_fsdp_core.py` to implement this interface. Further unification under this interface is coming in follow-up PRs.


[ghstack-poisoned]

awgu commented:
@pytorchbot merge
